specs/ChainSpec/Consensus.md

```diff
@@ -0,0 +1,146 @@
+```
+
+Where `send_approval` creates an approval with `prev_hash` and `prev_height` equal to `self.head_hash` and `self.head_height` correspondingly, `target_height` to the passed argument, and sends to the block producer at `target_height`.
+
+## Block Production
+We first define a convenience function to fetch approvals that can be included in a block at particular height:
+
+```python
+def get_approvals(self, target_height):
+    return [approval for approval
+                     in self.approvals
+                     if approval.target_height == target_height and
+                        approval.prev_height == self.head_height and
+                        (approval.prev_height + 1 != target_height or
+                         approval.prev_hash == self.head_hash)]
+```
+
+A block producer assigned for a particular height produces a block at that height whenever they have `get_approvals` return approvals from block producers whose stake collectively exceeds 2/3 of the total stake.
+
+## Safety
+
+We call approvals for which `prev_height + 1 = target_height` endorsements, and all other approvals skips.
+
+Note that with the implementation above a honest block producer can never produce two endorsements with the same `prev_height` (call this condition *conflicting endorsements*), neither can they produce a skip message `s` and an endorsement `e` such that `s.prev_height > e.prev_height and s.target_height <= e.target_height` (call this condition *conflicting skip and endorsement*).
+
+Say a block `B0` is final, i.e. there are blocks `B1` and `B2` that build on top of it and have sequential heights.
+
+Let's show that there's no other block `Bx` that has height higher than `B0` and doesn't have `B0` in its ancestry. Say it is not the case and such block `Bx` exists. Say `Bx` has the smallest height among such blocks. Thus, its previous block `Bp` has height less or equal to the height of `B0`. Consider three cases:
+
+1. `Bp.height < B0.height`. Then each approval in `Bx` is a skip. For each block producer that included a skip in `Bx` and an endorsement in `B1` these two messages conflict, thus at least 1/3 of stake-weighted block producers deviated from the protocol.
+2. `Bp.height == B0.height and Bx.height == B1.height`. Then each approval in `B1` is an endorsement. For each block producer that included an endorsement in `Bx` and an endorsement in `B1` these two messages conflict, thus at least 1/3 of stake-weighted block producers deviated from the protocol.
+3. `Bp.height == B0.height and Bx.height > B1.height`. Then each approval in `Bx` is a skip. For each block producer that included a skip in `Bx` and an endorsement in `B2` these two messages conflict, thus at least 1/3 of stake-weighted block producers deviated from the protocol.
+
+Thus, for any block with height higher than `B0.height` that doesn't have `B0` in its ancestry to be produced at least 1/3 of stake-weighted block producers must deviate from the protocol.
```

Comment:
I think that we should use some different term that "deviate from the protocol". Not every deviation from the protocol is slashable, and it is important that these particular deviations are.